nixops/backends/hetzner.py

```diff
@@ -650,7 +651,9 @@ def _wait_stop(self):
         """
         self.log_start("waiting for system to shutdown... ")
         dotlog = lambda: self.log_continue(".")  # NOQA
-        wait_for_tcp_port(self.main_ipv4, 22, open=False, callback=dotlog)
+        while self.try_ssh():
```

Comment:
This is a bit racy, because the timeout for `run_command` is only the SSH connect timeout, so if the machine is no longer reachable after the key exchange `try_ssh` could hang for minutes.

Btw. the same problem could occur in https://github.com/NixOS/nixops/pull/857, hence cc @nh2.

Reply:
If we'd switch to Python 3.x (I think it's long overdue), we could use [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) with the `timeout` argument for `logged_exec` and in `SSHMaster`, but in the interim it's still possible in Python 3.4 and lower using [`wait`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait) and [`kill`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.kill). We already have a `timeout` argument, so I'd either rename the argument or add another `command_timeout` argument. Note that we can't simply get rid of the connect timeout and just use the command timeout, because there might be long-running operations, like switching to a new configuration.